Issue #364 Add Support to Kafka Source for Resource spec

lberk opened the pull request:
As mentioned in Issue #364 , this is experimenting adding resource spec for 
* `resources.limits.cpu`
* `resources.limits.memory`
* `resources.requests.cpu`
* `resources.requests.memory`

As part of this change, we've moved the cmp test to kmp (was hitting errors similar to those seen in Serving 2622), Matching the approach in the serving repo.

matzew commented:
/ok-to-test

matzew commented:
In the `config/300-kafkasource.yaml`, let's add this:
```
            resources:
              properties:
                limits:
                  properties:
                    cpu:
                      type: string
                    memory:
                      type: string
                  type: object
                requests:
                  properties:
                    cpu:
                      type: string
                    memory:
                      type: string
                  type: object
              type: object
```

you can generate that file (but CAUTION) with:
```
cd contrib/kafka
go run ../../vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go all 
mv config/crds/sources_v1alpha1_kafkasource.yaml config/300-kafkasource.yaml
```
but only ... pick the relevant diff ... 


I'd also suggest adding this new feature to the samples, to make it a little bit more visible

matzew commented:
Regarding the coverage - let's add a test that does **NOT** specify the `Resource Limits`, so that we can test if the actual defaults are applied:

```
							Resources: corev1.ResourceRequirements{
								Limits: corev1.ResourceList{
									corev1.ResourceCPU:    resource.MustParse("100m"),
									corev1.ResourceMemory: resource.MustParse("500Mi"),
								},
								Requests: corev1.ResourceList{
									corev1.ResourceCPU:    resource.MustParse("10m"),
									corev1.ResourceMemory: resource.MustParse("10Mi"),
								},
							},
```

than we get back to 100% :)

matzew commented:
I deployed a source, no specific setting (assuming defaults), doing `k get kafkasources kafka-source -oyaml` gives me:

```
apiVersion: sources.eventing.knative.dev/v1alpha1
kind: KafkaSource
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"sources.eventing.knative.dev/v1alpha1","kind":"KafkaSource","metadata":{"annotations":{},"name":"kafka-source","namespace":"default"},"spec":{"bootstrapServers":"my-cluster-kafka-bootstrap.kafka:9092","consumerGroup":"my-group","sink":{"apiVersion":"serving.knative.dev/v1alpha1","kind":"Service","name":"event-display"},"topics":"my-topic"}}
  creationTimestamp: "2019-05-07T09:11:12Z"
  generation: 2
  name: kafka-source
  namespace: default
  resourceVersion: "21585"
  selfLink: /apis/sources.eventing.knative.dev/v1alpha1/namespaces/default/kafkasources/kafka-source
  uid: 0f203a34-70a8-11e9-9cdb-e452b3720906
spec:
  bootstrapServers: my-cluster-kafka-bootstrap.kafka:9092
  consumerGroup: my-group
  net:
    sasl:
      password: {}
      user: {}
    tls: {}
  resources:
    limits: {}
    requests: {}
  sink:
    apiVersion: serving.knative.dev/v1alpha1
    kind: Service
    name: event-display
  topics: my-topic
```

Expected ?

matzew commented:
looking in the underlying `deployment`, I do indeed see the defaults:

```
...
        resources:
          limits:
            cpu: 100m
            memory: 100M
          requests:
            cpu: 10m
            memory: 10M
...
```

matzew commented:
Now,  updating/changing a _running_ source (by explicitly adding **NON-DEFAULT** settings) and runningk `get kafkasources kafka-source -oyaml`

gives me:

```
apiVersion: sources.eventing.knative.dev/v1alpha1
kind: KafkaSource
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"sources.eventing.knative.dev/v1alpha1","kind":"KafkaSource","metadata":{"annotations":{},"name":"kafka-source","namespace":"default"},"spec":{"bootstrapServers":"my-cluster-kafka-bootstrap.kafka:9092","consumerGroup":"my-group","resources":{"limits":{"cpu":"200m","memory":"200M"},"requests":{"cpu":"20m","memory":"20M"}},"sink":{"apiVersion":"serving.knative.dev/v1alpha1","kind":"Service","name":"event-display"},"topics":"my-topic"}}
  creationTimestamp: "2019-05-07T09:11:12Z"
  generation: 3
  name: kafka-source
  namespace: default
  resourceVersion: "22294"
  selfLink: /apis/sources.eventing.knative.dev/v1alpha1/namespaces/default/kafkasources/kafka-source
  uid: 0f203a34-70a8-11e9-9cdb-e452b3720906
spec:
  bootstrapServers: my-cluster-kafka-bootstrap.kafka:9092
  consumerGroup: my-group
  net:
    sasl:
      password: {}
      user: {}
    tls: {}
  resources:
    limits:
      cpu: 200m
      memory: 200M
    requests:
      cpu: 20m
      memory: 20M
  sink:
```

But looking at the underlying `deployment` I still see:
```
        resources:
          limits:
            cpu: 100m
            memory: 100M
          requests:
            cpu: 10m
            memory: 10M

```

matzew commented:
Now, creating a new source, where I set values, like:

```
  resources:
    limits:
      cpu: 200m
      memory: 200Mi
    requests:
      cpu: 200m
      memory: 200Mi

```

It's all correct in the `kafkasource` AND the underlying deployment

matzew commented on contrib/kafka/pkg/reconciler/resources/receive_adapter.go:
```diff
@@ -80,6 +81,33 @@ func MakeReceiveAdapter(args *ReceiveAdapterArgs) *v1.Deployment {
+	}
+	RequestResourceMemory, err := resource.ParseQuantity(args.Source.Spec.Resources.Requests.ResourceMemory)
+	if err != nil {
+		RequestResourceMemory = resource.MustParse("100M")
```
`100Mi` ?

matzew commented on contrib/kafka/pkg/reconciler/resources/receive_adapter.go:
```diff
@@ -80,6 +81,33 @@ func MakeReceiveAdapter(args *ReceiveAdapterArgs) *v1.Deployment {
+	}
+	LimitResourceMemory, err := resource.ParseQuantity(args.Source.Spec.Resources.Limits.ResourceMemory)
+	if err != nil {
+		LimitResourceMemory = resource.MustParse("10M")
```
`100Mi` ?

matzew commented on contrib/kafka/pkg/reconciler/resources/receive_adapter.go:
```diff
@@ -80,6 +81,33 @@ func MakeReceiveAdapter(args *ReceiveAdapterArgs) *v1.Deployment {
 	}
+	RequestResourceCPU, err := resource.ParseQuantity(args.Source.Spec.Resources.Requests.ResourceCPU)
+	if err != nil {
+		RequestResourceCPU = resource.MustParse("100m")
```
btw. must be less than or equal to cpu limit

matzew replied:
see: https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/#specify-a-cpu-request-and-a-cpu-limit

matzew commented:
For the defaults, lets use:

requests": {"memory": "512Mi", "cpu": "250m"}, "limits": {"memory":
"512Mi", "cpu": "250m"}}


we use this as our setting currently


On Tue 7. May 2019 at 07:23, covbot <notifications@github.com> wrote:

> The following is the coverage report on pkg/.
> Say /test pull-knative-eventing-sources-go-coverage to re-run this
> coverage report
> File Old Coverage New Coverage Delta
> contrib/kafka/pkg/reconciler/resources/receive_adapter.go
> <https://storage.cloud.google.com/knative-prow/pr-logs/pull/knative_eventing-sources/399/pull-knative-eventing-sources-go-coverage/1125631815193726979/artifacts/line-cov.html#file1>
> 100.0% 80.0% -20.0
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/knative/eventing-sources/pull/399#issuecomment-489913099>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABGPTSJQRDZELJPIHQ2T6TPUEG6XANCNFSM4HLEMG7Q>
> .
>
-- 
Sent from Gmail Mobile

lberk commented:
Updated as requested, and rebased on master
* config/300-kafkasource.yaml - add in resource properties
* pkg/reconcilier/resources/receive_adapter.go - increase the defaults to proper values
* pkg/reconcilier/resources/receive_adapter_test.go - add second testcase (we should probably create a proper test struct in the future)
* samples/event-sources.yaml - add in an example resource specification

matzew commented:
deploying a source, without the `resource` config, I still see it empty, when running ` k get kafkasources kafka-source -oyaml` I am getting:

```
...
spec:
  bootstrapServers: my-cluster-kafka-bootstrap.kafka:9092
  consumerGroup: new
  net:
    sasl:
      password: {}
      user: {}
    tls: {}
  resources:
    limits: {}
    requests: {}
  sink:
    apiVersion: serving.knative.dev/v1alpha1
    kind: Service
    name: event-display
  topics: my-topic
```

I'd expect the `sepc.resources` would tell the actual defaults

mattmoor-sockpuppet commented on contrib/kafka/pkg/apis/sources/v1alpha1/kafka_types.go:
```diff
@@ -105,6 +120,9 @@ type KafkaSourceSpec struct {
 	// Adapter Deployment.
 	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// Resource limits and Reqeust specifications of the Recieve Adapter Deployment
```
<!--typo-bot-->
```suggestion
	// Resource limits and Reqeust specifications of the Receive Adapter Deployment
```

mattmoor-sockpuppet commented on contrib/kafka/pkg/apis/sources/v1alpha1/kafka_types.go:
```diff
@@ -105,6 +120,9 @@ type KafkaSourceSpec struct {
 	// Adapter Deployment.
 	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// Resource limits and Request specifications of the Recieve Adapter Deployment
```
<!--typo-bot-->
```suggestion
	// Resource limits and Request specifications of the Receive Adapter Deployment
```

matzew commented:
/lgtm
/approve

Harwayne commented on contrib/kafka/pkg/reconciler/resources/receive_adapter.go:
```diff
@@ -73,11 +74,38 @@ func MakeReceiveAdapter(args *ReceiveAdapterArgs) *v1.Deployment {
 
+	RequestResourceCPU, err := resource.ParseQuantity(args.Source.Spec.Resources.Requests.ResourceCPU)
+	if err != nil {
+		RequestResourceCPU = resource.MustParse("250m")
```
Should this be an error instead of silently ignoring the value and choosing a default? For example, if my spec has `5000n` (when I intended `5000m`), would I expect the KafkaSource to have a failing status or to create a receive adapter with 250m?

lberk replied:
@Harwayne I tried implementing this (as per our discussion yesterday on slack), but I'm finding that when I specify an invalid resource amount (even after implementing a corresponding `KafkaStatus*` status functions) it seems that the controller fails to even run to the point where the Status functions would be triggered.  My best guess is that perhaps the admission webhook? might be already verifying the validity of the deployment's `corev1.ResourceRequirements` ?

```
ERROR: logging before flag.Parse: E0524 15:06:51.020134       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.KafkaSource: v1alpha1.KafkaSourceList.Items: []v1alpha1.KafkaSource: v1alpha1.KafkaSource.Spec: v1alpha1.KafkaSourceSpec.Sink: Resources: v1.ResourceRequirements.Requests: unmarshalerDecoder: quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$', error found in #10 byte of ...|pu":"500p"}},"sink":|..., bigger context ...|},"key":{}}},"resources":{"requests":{"cpu":"500p"}},"sink":{"apiVersion":"serving.knative.dev/v1alp|...
```
Is there another approach you'd like to explore?  Or, with the change to `corev1.ResourceRequirements` can we consider this point resolved?

Harwayne replied:
Was that failing CO created before or after the change to `corev1.ResourceRequirements`? If it was before, then it might be something that the webhook/API server would correctly handle for us now. If it was after, then that is a big problem and we might need to revert to your original implementation.

lberk replied:
CO?  I submitted that (erroneous) `resource.yaml` file (which lead to that error message) after the change to `corev1.ResourceRequirements`

Harwayne replied:
CO = Custom Object. An actual instance of a CRD. For example, Broker = CRD, the Broker named 'default' in the namespace 'default' = CO.

Was that log from the webhook or the controller?

lberk replied:
Gotcha, thanks!  the output was produced via controller logs;
`k -n knative-sources logs kafka-controller-manager-0`

Harwayne replied:
I don't know why it is failing in this way, but this not a failure we can tolerate, so we may need to revert to your earlier implementation of hand rolling the equivalent of `corev1.ResourceRequirements`. 

@n3wscott @vaikas-google any ideas why this is failing in this way? Namely that the API server seems to accept something that can't be unmarshaled into the Go object.

n3wscott replied:
you are setting `quantities` to `"500p"` and the regex is `'^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'`

try changing it to `"500P"` ?

lberk replied:
Thanks @n3wscott while that works, I was under the impression that wasn't what Adam was asking about.  Perhaps I've misunderstood @Harwayne 's initial review comment.  Are we trying to handle;
1. The case where the resourcespec itself is incorrect (ie, 500p -- lowercase p) OR
2. The case where the underlying resourcespec matches the correct syntax, but is not appropriate for the platform its running on? (ie, 500G on my laptop)?

I thought it was the former?

Harwayne replied:
This initial comment thread was about what happens when the value is invalid. The initial code silently replaced it with a default value. I argued it should set a status error message instead.

Another thread recommended replacing the hand rolled resource requirements with the actual `corev1.ResourceRequirements`, but that seemed to cause weird failures _only_ in the controller, which really scares me. So if we can't find the cause of that (such that bad values are rejected either before getting to the controller, or the controller handles them gracefully), then I said we should move back to the hand rolled resource requirements.

lberk replied:
@Harwayne thanks for the clarification.  Unless @n3wscott (or anybody else) suggests otherwise, I'm happy to start reverting to my previous implementation.

Harwayne replied:
Now that we have reverted, I think my original question still stands:

> Should this be an error instead of silently ignoring the value and choosing a default? For example, if my spec has 5000n (when I intended 5000m), would I expect the KafkaSource to have a failing status or to create a receive adapter with 250m?

Essentially, can this function return `(*v1.Deployment, error)` rather than silently ignoring bad values?

Harwayne replied:
I understand now. Before this is called, `checkResourcesStatus(src)` is called which verifies the fields. 

Perhaps change these to panics, rather than defaulting? We expect them to never happen, and I would rather they fail loudly if they do. However, I'll leave that decision to you.

Harwayne commented on contrib/kafka/pkg/apis/sources/v1alpha1/kafka_types.go:
```diff
@@ -105,6 +120,9 @@ type KafkaSourceSpec struct {
 	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// Resource limits and Request specifications of the Receive Adapter Deployment
+	Resources KafkaResourceSpec `json:"resources,omitempty"`
```
Why not corev1.ResourceRequirements?

lberk replied:
Updated in 37df1a3

danp replied:
I was going to leave the same feedback but saw this, that it was done, and then that the change was reverted. Was there trouble with using corev1.ResourceRequirements?

lberk replied:
@danp please see the conversation in https://github.com/knative/eventing-sources/pull/399#discussion_r288702244
Short answer, yes, it led to error messages only visible within the controller logs

mattmoor-sockpuppet commented on contrib/kafka/pkg/reconciler/kafkasource.go:
```diff
@@ -122,14 +124,64 @@ func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) error
+		key:   "Limit.Memory",
+		field: src.Spec.Resources.Limits.ResourceMemory,
+	}} {
+		// In the event the field isn't specified, we assign a default in the recieve_adapter
```
<!--typo-bot-->
```suggestion
		// In the event the field isn't specified, we assign a default in the receive_adapter
```

matzew commented:
/approve

matzew commented:
@n3wscott or @grantr   This PR was `/lgtm`'d by @Harwayne, and I could, in theory, approve it.

However, since it also updates non Kafka file (`Gopkg.lock`), my /approve is not good enough :-(
